File: tape/sdk/java/src/test/java/dev/tape/ReactionsTest.java
package dev.tape;

import dev.tape.proto.*;
import org.junit.jupiter.api.*;

import java.io.File;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.nio.file.*;
import java.time.Duration;
import java.util.Iterator;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

import static org.junit.jupiter.api.Assertions.*;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

/** Reactions / tasks / SubscribeBySubject round-trip against the live Rust server.
 *
 * <p>The Python equivalents use a function-scoped fixture (one fresh server per
 * test). We mirror that here — accumulating reactions across tests confuses the
 * matcher's batched-cursor traversal under load. */
public class ReactionsTest {

    Process server;
    String url;
    int port;

    static Path serverBin;

    @BeforeAll
    static void checkBinary() {
        serverBin = Paths.get("../../server/target/debug/tape-server").toAbsolutePath().normalize();
        assumeTrue(Files.exists(serverBin), "tape-server not built (run `cargo build` in tape/server)");
    }

    @BeforeEach
    void start() throws Exception {
        Reactions.clearRegistry();
        try (java.net.ServerSocket s = new java.net.ServerSocket(0, 1,
                java.net.InetAddress.getByName("127.0.0.1"))) {
            port = s.getLocalPort();
        }
        ProcessBuilder pb = new ProcessBuilder(serverBin.toString(),
                "--listen", "127.0.0.1:" + port, "--store", "memory")
                .redirectOutput(new File("/dev/null"))
                .redirectError(new File("/dev/null"));
        pb.environment().put("RUST_LOG", "tape_server=warn");
        server = pb.start();
        url = "tape://127.0.0.1:" + port;
        waitFor("127.0.0.1", port, 15_000);
    }

    @AfterEach
    void stop() throws InterruptedException {
        if (server != null) {
            server.destroy();
            server.waitFor(5, TimeUnit.SECONDS);
        }
    }

    static void waitFor(String host, int port, int timeoutMs) {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            try (Socket s = new Socket()) { s.connect(new InetSocketAddress(host, port), 1000); return; }
            catch (IOException ignored) {}
            try { Thread.sleep(100); } catch (InterruptedException e) { return; }
        }
        fail("server never came up at " + host + ":" + port);
    }

    /** Block until {@code probe} returns a non-null/non-empty value, or fail. */
    static <T> T waitFor(Supplier<T> probe, long timeoutMs) {
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline) {
            try {
                T t = probe.get();
                if (t == null) { /* keep polling */ }
                else if (t instanceof java.util.Collection && ((java.util.Collection<?>) t).isEmpty()) { /* keep */ }
                else return t;
            } catch (Throwable ignored) {}
            try { Thread.sleep(50); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); return null; }
        }
        return null;
    }

    // ── 1. New client methods round-trip via the live server ────────────────

    @Test
    void clientReactionRpcsRoundTrip() {
        try (TapeClient c = new TapeClient(url)) {
            String stableId = "java-rt-" + UUID.randomUUID().toString().substring(0, 6);
            Reaction r = c.registerReaction(new RegisterReactionOpts()
                    .reactionId(stableId)
                    .name("client-roundtrip")
                    .subjectPattern("/tape/value/changed/java-rt/**")
                    .handlerKind(HandlerKind.HANDLER_KIND_TASK)
                    .maxConcurrency(2));
            assertEquals(stableId, r.getReactionId());
            assertEquals("/tape/value/changed/java-rt/**", r.getSubjectPattern());
            assertEquals(HandlerKind.HANDLER_KIND_TASK, r.getHandlerKind());

            // ListReactions: filter by pattern prefix.
            List<Reaction> all = c.listReactions("");
            assertTrue(all.stream().anyMatch(x -> x.getReactionId().equals(stableId)),
                    "list_reactions('') should include our reaction");
            List<Reaction> filtered = c.listReactions("/tape/value/changed/java-rt/**");
            assertTrue(filtered.stream().anyMatch(x -> x.getReactionId().equals(stableId)),
                    "list_reactions(pattern) should still include our reaction");

            // Trigger a match, then claim + complete the resulting task.
            c.pb().writeValue(WriteValueRequest.newBuilder()
                    .setNamespace("java-rt").setKey("k1").setValueJson("1").setIfVersion(-1).setWriter("t").build());

            List<Task> tasks = waitFor(
                    () -> c.listTasks(stableId, TaskStatus.TASK_STATUS_UNSPECIFIED, 100),
                    5_000);
            assertNotNull(tasks, "matcher should have created a task");
            assertFalse(tasks.isEmpty(), "expected >=1 task");

            String owner = "java-owner";
            List<Task> claimed = c.claimTasks(new ClaimTasksOpts()
                    .reactionId(stableId).owner(owner).max(10));
            assertFalse(claimed.isEmpty(), "claim should return >=1 task");
            String taskId = claimed.get(0).getTaskId();

            Task done = c.completeTask(taskId, owner);
            assertEquals(TaskStatus.TASK_STATUS_DONE, done.getStatus());

            // Deregister.
            assertTrue(c.deregisterReaction(stableId), "deregister should return true");
            assertFalse(c.listReactions("").stream().anyMatch(x ->
                    x.getReactionId().equals(stableId) && !x.getDeleted()),
                    "post-deregister, the reaction must not appear as live");
        }
    }

    // ── 2. A task reaction's handler runs via runDispatcher ────────────────

    @Test
    void dispatcherRunsTaskHandlerAndCompletesTask() throws Exception {
        CopyOnWriteArrayList<String> seenSubjects = new CopyOnWriteArrayList<>();

        Reactions.ReactionDef def = new Reactions.ReactionDef();
        def.subjectPattern = "/tape/value/changed/java-disp/**";
        def.name = "java-disp-handler";
        def.maxConcurrency = 2;
        def.handler = env -> seenSubjects.add(env.subject());
        Reactions.on(def);

        try (TapeClient c = new TapeClient(url)) {
            List<Reaction> rs = Reactions.registerAll(c, "t" + UUID.randomUUID().toString().substring(0, 4) + "-");
            assertEquals(1, rs.size());
            String rid = rs.get(0).getReactionId();

            // Trigger.
            c.pb().writeValue(WriteValueRequest.newBuilder()
                    .setNamespace("java-disp").setKey("k1").setValueJson("{\"v\":1}").setIfVersion(-1).setWriter("t").build());

            // Wait for the matcher to create a task.
            List<Task> pending = waitFor(
                    () -> c.listTasks(rid, TaskStatus.TASK_STATUS_UNSPECIFIED, 10), 5_000);
            assertNotNull(pending, "task should exist");

            // Run one dispatcher pass.
            Reactions.runDispatcher(c, new Reactions.RunDispatcherOpts()
                    .once(true).register(false).pollInterval(Duration.ofMillis(50)));

            // Wait for the handler to have run.
            for (int i = 0; i < 80 && seenSubjects.isEmpty(); i++) Thread.sleep(50);
            assertFalse(seenSubjects.isEmpty(), "handler must have been invoked");
            assertTrue(seenSubjects.get(0).startsWith("/tape/value/changed/java-disp/"),
                    "subject must match pattern; got " + seenSubjects.get(0));

            // Task should be DONE.
            List<Task> done = waitFor(
                    () -> c.listTasks(rid, TaskStatus.TASK_STATUS_DONE, 10), 5_000);
            assertNotNull(done, "task should reach DONE");
            assertFalse(done.isEmpty());
        }
    }

    // ── 3. bootstrap_from_head=true skips the backlog ──────────────────────

    @Test
    void bootstrapFromHeadSkipsBacklog() throws Exception {
        // What this verifies: when a reaction is registered with
        // `bootstrap_from_head=true` AFTER N journal entries already exist,
        // those N entries must NOT produce tasks — the cursor is seeded at
        // head, so the reaction only sees what happens going forward. We
        // assert the backlog-skip contract (the post-registration delivery
        // path is exercised by `dispatcherRunsTaskHandlerAndCompletesTask`
        // and `clientReactionRpcsRoundTrip`, which both register reactions
        // with `bootstrap_from_head=false`).
        String ns = "java-boot-" + UUID.randomUUID().toString().substring(0, 6);
        try (TapeClient c = new TapeClient(url)) {
            // Pre-write some entries — these form the backlog the reaction will skip.
            for (int i = 0; i < 5; i++) {
                c.pb().writeValue(WriteValueRequest.newBuilder()
                        .setNamespace(ns).setKey("k").setValueJson(String.valueOf(i)).setIfVersion(-1).setWriter("t").build());
            }

            String rid = "java-boot-r-" + UUID.randomUUID().toString().substring(0, 6);
            Reaction r = c.registerReaction(new RegisterReactionOpts()
                    .reactionId(rid)
                    .name("bootstrap-head")
                    .subjectPattern("/tape/value/changed/" + ns + "/**")
                    .handlerKind(HandlerKind.HANDLER_KIND_TASK)
                    .bootstrapFromHead(true));
            assertEquals(rid, r.getReactionId());
            // bootstrap_from_head is a registration-time intent (seeds the
            // per-shard cursor at head); the server intentionally does NOT
            // echo the flag back on the persisted Reaction row.

            // Give the matcher a moment; with bootstrap_from_head=true it must NOT
            // pick up the 5 pre-existing entries.
            Thread.sleep(1_500);
            List<Task> backlog = c.listTasks(rid, TaskStatus.TASK_STATUS_UNSPECIFIED, 100);
            assertEquals(0, backlog.size(),
                    "bootstrap_from_head must skip pre-existing entries; got: "
                            + backlog.stream().map(Task::getSubject).toList());
        }
    }

    // ── 4. subscribeBySubject filters by pattern ───────────────────────────

    @Test
    void subscribeBySubjectFiltersByPattern() throws Exception {
        // We drive matching journal entries via `RecordDecision` (each
        // decision becomes a /tape/decision/recorded/<run_id>/<idx> row).
        // Avoiding `WriteValue` here: every value journal entry currently
        // pins (run_id="", seq=0), so a second value write silently fails
        // the journal insert — making it unusable for "stream multiple
        // matching events" tests. Decisions are per-run-sequenced, so they
        // don't collide.
        String app = "java-sub-" + UUID.randomUUID().toString().substring(0, 6);
        try (TapeClient writer = new TapeClient(url);
             TapeClient reader = new TapeClient(url)) {

            // Run 1 + two decisions → two matching journal entries.
            BeginRunResponse r1 = writer.beginRun(app, "u", "s", "inv-1", "owner", 60_000);
            writer.recordDecision(r1.getRunId(), 0, "m", "{}", "{}", "", "p1");
            writer.recordDecision(r1.getRunId(), 1, "m", "{}", "{}", "", "p1");
            // Run 2 + one decision under a DIFFERENT app → non-matching
            // subject (pattern is scoped to the first run's run_id).
            BeginRunResponse r2 = writer.beginRun(app + "-other", "u", "s", "inv-2", "owner", 60_000);
            writer.recordDecision(r2.getRunId(), 0, "m", "{}", "{}", "", "p1");

            // Pattern matches decisions for run_id=r1 only.
            String pattern = "/tape/decision/recorded/" + r1.getRunId() + "/**";

            CopyOnWriteArrayList<EventEntry> hits = new CopyOnWriteArrayList<>();
            Thread t = new Thread(() -> {
                try {
                    java.util.Iterator<EventEntry> it = reader.subscribeBySubject(pattern, "", 0);
                    while (it.hasNext() && hits.size() < 5) {
                        hits.add(it.next());
                    }
                } catch (Throwable ignored) {
                    // CANCELLED on cleanup is expected.
                }
            }, "sub-reader");
            t.setDaemon(true);
            t.start();

            long deadline = System.currentTimeMillis() + 8_000;
            while (System.currentTimeMillis() < deadline && hits.size() < 2) {
                Thread.sleep(50);
            }
            // Close the reader's channel — that ends the streaming RPC.
            reader.close();
            assertTrue(hits.size() >= 2, "expected >=2 matching events, got " + hits.size()
                    + " (subjects=" + hits.stream().map(EventEntry::getSubject).toList() + ")");
            for (EventEntry e : hits) {
                assertTrue(e.getSubject().startsWith("/tape/decision/recorded/" + r1.getRunId() + "/"),
                        "subject must match pattern; got " + e.getSubject());
            }
        }
    }

    // ── 5. Debouncer / TokenBucket primitives (unit) ───────────────────────

    @Test
    void debouncerCoalescesSameSubject() {
        Reactions.Debouncer d = new Reactions.Debouncer(200);
        assertTrue(d.allow("s/x"));
        assertFalse(d.allow("s/x"), "second hit inside window must be coalesced");
        assertTrue(d.allow("s/y"), "different subject must pass");
    }

    @Test
    void tokenBucketAllowsBurst() throws Exception {
        Reactions.TokenBucket tb = new Reactions.TokenBucket(10);
        // Rate=10 means starting bucket is full (10 tokens). Burst acquire 10
        // should complete near-instantly.
        long t0 = System.nanoTime();
        for (int i = 0; i < 10; i++) tb.acquire();
        long elapsedMs = (System.nanoTime() - t0) / 1_000_000;
        assertTrue(elapsedMs < 500, "burst of 10 should complete fast (got " + elapsedMs + "ms)");
    }

    @Test
    void parseJsonObjectHandlesBasicCases() {
        assertNull(Reactions.parseJsonObject(null));
        assertNull(Reactions.parseJsonObject(""));
        assertNull(Reactions.parseJsonObject("\"just-a-string\""));
        java.util.Map<String, Object> m = Reactions.parseJsonObject(
                "{\"a\":1,\"b\":\"x\",\"c\":true,\"d\":null}");
        assertNotNull(m);
        assertEquals(1L, m.get("a"));
        assertEquals("x", m.get("b"));
        assertEquals(Boolean.TRUE, m.get("c"));
        assertNull(m.get("d"));
    }
}

File: tape/sdk/typescript/test/reactions.test.ts
// End-to-end tests for the event-bus surface (registerReaction, ClaimTasks,
// CompleteTask, NackTask, ListTasks, SubscribeBySubject) and the in-proc
// dispatcher (`runDispatcher` / `on*` helpers). Mirrors the high-level
// coverage of `tape/tests/test_event_bus.py`.
//
// Each test spawns a fresh tape-server (matching the Python suite's
// function-scoped `tape_server` fixture) — the server's matcher tails the
// journal across all reactions, and a per-test fresh server gives the
// deterministic isolation reactions need.
//
// Skips if the Rust tape-server binary isn't built (same pattern as
// client.test.ts).

import test from 'node:test';
import assert from 'node:assert/strict';
import { spawn, type ChildProcess } from 'node:child_process';
import { existsSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import * as net from 'node:net';
import {
  TapeClient, HandlerKind, TaskStatus,
  on, registerAll, runDispatcher, _clearRegistry,
} from '../src/index.ts';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SERVER_BIN = join(__dirname, '..', '..', '..', 'server', 'target', 'debug', 'tape-server');
const SERVER_AVAILABLE = existsSync(SERVER_BIN);

function freePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const s = net.createServer();
    s.listen(0, '127.0.0.1', () => {
      const p = (s.address() as net.AddressInfo).port;
      s.close(() => resolve(p));
    });
    s.on('error', reject);
  });
}

async function waitFor(host: string, port: number, timeoutMs = 15_000): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const ok = await new Promise<boolean>((r) => {
      const s = net.connect(port, host);
      s.on('connect', () => { s.end(); r(true); });
      s.on('error', () => { r(false); });
    });
    if (ok) return;
    await new Promise((r) => setTimeout(r, 100));
  }
  throw new Error(`server never came up at ${host}:${port}`);
}

async function waitUntil<T>(
  fn: () => Promise<T> | T,
  timeoutMs = 5000,
  intervalMs = 100,
): Promise<T | null> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const v = await fn();
    if (v) return v;
    await new Promise((r) => setTimeout(r, intervalMs));
  }
  return null;
}

interface Server { proc: ChildProcess; url: string; }

async function spawnServer(): Promise<Server> {
  const port = await freePort();
  const proc = spawn(SERVER_BIN, ['--listen', `127.0.0.1:${port}`, '--store', 'memory'], {
    stdio: 'ignore',
    env: { ...process.env, RUST_LOG: 'tape_server=warn' },
  });
  await waitFor('127.0.0.1', port);
  return { proc, url: `tape://127.0.0.1:${port}` };
}

function killServer(s: Server): void {
  try { s.proc.kill('SIGTERM'); } catch { /* already dead */ }
}

// Wrap a test body that needs a fresh server. Skips cleanly if the binary
// isn't built.
function withServer(name: string, body: (s: Server) => Promise<void>): void {
  test(name, async (t) => {
    if (!SERVER_AVAILABLE) { t.skip(`tape-server not built at ${SERVER_BIN}`); return; }
    _clearRegistry();
    const s = await spawnServer();
    try { await body(s); }
    finally { _clearRegistry(); killServer(s); }
  });
}


// ── 1. Journal entries carry global_seq, subject, schema_version ───────────

withServer('journal entries have non-zero global_seq and non-empty subject',
  async ({ url }) => {
    const c = new TapeClient(url);
    try {
      const begun: any = await c.beginRun({
        appName: 'a', userId: 'u', sessionId: 's-gs',
        invocationId: `inv-${Date.now().toString(36)}`,
        leaseOwner: 'test', leaseTtlMs: 60_000,
      });
      await c.recordDecision({ runId: begun.runId, decisionIndex: 0,
        responseJson: '{"plan":1}' });
      await c.writeValue({ namespace: 'treasury', key: 'fx_rate',
        valueJson: '"1.10"', writer: 'oracle' });

      // give the server time to drain its in-process notify
      await new Promise((r) => setTimeout(r, 200));

      // Drain /tape/** with a short deadline. The server's stream surfaces
      // DEADLINE_EXCEEDED once the deadline passes, which the client treats
      // as a clean end-of-stream.
      const entries: any[] = [];
      for await (const e of c.subscribeBySubject({
        subjectPattern: '/tape/**', fromGlobalSeq: 0, timeoutMs: 800,
      }) as any) {
        entries.push(e);
      }

      assert.ok(entries.length >= 1, `expected at least one /tape/ entry, got ${entries.length}`);
      for (const e of entries) {
        assert.ok(Number(e.globalSeq) > 0, `global_seq must be > 0; got ${e.globalSeq}`);
        assert.ok(String(e.subject).startsWith('/tape/'),
          `subject must start with /tape/; got ${e.subject}`);
        assert.equal(Number(e.schemaVersion), 1,
          `schema_version must default to 1; got ${e.schemaVersion}`);
      }
      const seqs = entries.map((e) => Number(e.globalSeq));
      const sortedSeqs = [...seqs].sort((a, b) => a - b);
      assert.deepEqual(seqs, sortedSeqs, 'global_seq must be monotonic');
    } finally { try { c.close(); } catch { /* */ } }
  });


// ── 2. subscribeBySubject only returns matching entries ────────────────────

withServer('subscribeBySubject only returns matching entries', async ({ url }) => {
  const c = new TapeClient(url);
  try {
    await c.writeValue({ namespace: 'treasury', key: 'k1',
      valueJson: '1', writer: 'tt' });
    await c.writeValue({ namespace: 'other', key: 'k1',
      valueJson: '2', writer: 'tt' });
    await new Promise((r) => setTimeout(r, 200));

    const entries: any[] = [];
    for await (const e of c.subscribeBySubject({
      subjectPattern: '/tape/value/changed/treasury/**', fromGlobalSeq: 0, timeoutMs: 800,
    }) as any) {
      entries.push(e);
    }

    assert.ok(entries.length >= 1, 'expected ≥1 treasury entry');
    for (const e of entries) {
      assert.ok(e.subject.startsWith('/tape/value/changed/treasury/'),
        `got ${e.subject} — subject pattern should filter`);
    }
  } finally { try { c.close(); } catch { /* */ } }
});


// ── 3. Registered task reaction → dispatcher runs handler → complete_task ──

withServer('registered task reaction runs the handler via the dispatcher',
  async ({ url }) => {
    const called: any[] = [];
    on('/tape/value/changed/disp-test/**', (env) => { called.push(env); },
      { maxConcurrency: 2 });

    const c = new TapeClient(url);
    try {
      const rs = await registerAll({ url, prefix: `t${Date.now().toString(36)}-` });
      assert.equal(rs.length, 1, 'one registered reaction');
      const rid = rs[0].reactionId;
      assert.ok(rid, 'server-assigned reaction_id');

      await c.writeValue({ namespace: 'disp-test', key: 'k1',
        valueJson: '{"v":1}', writer: 't' });

      // wait for the matcher to enqueue a task
      const tasks: any = await waitUntil(async () => {
        const ts = await c.listTasks({ reactionId: rid, limit: 10 });
        return ts.length > 0 ? ts : null;
      }, 5000);
      assert.ok(tasks && tasks.length >= 1, `expected ≥1 pending task; got ${tasks?.length}`);

      // run dispatcher once — should claim, run, complete
      await runDispatcher({ url, once: true, register: false, pollIntervalMs: 50 });

      assert.ok(called.length > 0, 'handler should be invoked');
      assert.ok(called[0]?.task?.subject?.includes('disp-test'),
        `envelope.task.subject should contain disp-test; got ${called[0]?.task?.subject}`);
      // Payload is the journal entry's payload_json; for a value-write, the
      // server packages `{namespace, key, value: {namespace, key, value_json, version}, ...}`.
      assert.equal(called[0]?.payload?.namespace, 'disp-test');
      assert.equal(called[0]?.payload?.value?.value_json, '{"v":1}');

      // task should be DONE
      const done = await waitUntil(async () => {
        const ts = await c.listTasks({ reactionId: rid, status: TaskStatus.DONE, limit: 10 });
        return ts.length > 0 ? ts : null;
      }, 4000);
      assert.ok(done && done.length > 0, 'task should be DONE');
    } finally { c.close(); }
  });


// ── 4. bootstrap_from_head skips backlog ───────────────────────────────────

withServer('bootstrap_from_head skips the existing backlog', async ({ url }) => {
  const c = new TapeClient(url);
  try {
    // Write several pre-existing entries to a namespace.
    const ns = 'bootstrap-test';
    for (let i = 0; i < 3; i++) {
      await c.writeValue({ namespace: ns, key: `k${i}`,
        valueJson: `${i}`, writer: 't' });
    }
    await new Promise((r) => setTimeout(r, 200));

    // Register a reaction AFTER the backlog with bootstrap_from_head=true.
    // The flag must be propagated through the gRPC layer (TS SDK
    // responsibility) and honoured by the server's cursor-seed logic — the
    // reaction must not see any of the 3 pre-existing matching entries.
    const r: any = await c.registerReaction({
      name: 'after-only',
      subjectPattern: `/tape/value/changed/${ns}/**`,
      handlerKind: HandlerKind.TASK,
      bootstrapFromHead: true,
    });
    assert.ok(r.reactionId, 'server-assigned reaction_id');

    // Give the matcher a beat to (not) process the backlog.
    await new Promise((rr) => setTimeout(rr, 1000));

    const backlogTasks = await c.listTasks({ reactionId: r.reactionId, limit: 50 });
    assert.equal(backlogTasks.length, 0,
      `bootstrap_from_head should skip backlog; got ${backlogTasks.length} tasks`);
  } finally { c.close(); }
});


// ── 5. listReactions returns what we registered ────────────────────────────

withServer('listReactions returns registered reactions; deregisterReaction works',
  async ({ url }) => {
    const c = new TapeClient(url);
    try {
      const pattern = `/tape/effect/confirmed/list-test/**`;
      const r: any = await c.registerReaction({
        name: 'list-me',
        subjectPattern: pattern,
        handlerKind: HandlerKind.TASK,
      });
      const reactions = await c.listReactions(pattern);
      assert.ok(reactions.some((x: any) => x.reactionId === r.reactionId),
        'just-registered reaction should appear in listReactions');

      const ok = await c.deregisterReaction(r.reactionId);
      assert.equal(ok, true, 'deregisterReaction returns true on success');

      const after = await c.listReactions(pattern);
      assert.equal(after.length, 0,
        'deregistered reaction should not appear in listReactions');
    } finally { c.close(); }
  });
